test-data/unit/check-typeddict.test

```diff
@@ -2025,6 +2025,193 @@ v = {bad2: 2}  # E: Extra key "bad" for TypedDict "Value"
 [builtins fixtures/dict.pyi]
 [typing fixtures/typing-typeddict.pyi]
 
+[case testOperatorContainsNarrowsTypedDicts_unionWithList]
+from __future__ import annotations
+from typing import assert_type, TypedDict, Union
+from typing_extensions import final
+
+@final
+class D(TypedDict):
+    foo: int
+
+
+d: D | list[str]
+
+if 'foo' in d:
+    assert_type(d, Union[D, list[str]])
+else:
+    assert_type(d, list[str])
+
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-typeddict.pyi]
+
+[case testOperatorContainsNarrowsTypedDicts_total]
+from __future__ import annotations
+from typing import assert_type, Literal, TypedDict, TypeVar, Union
+from typing_extensions import final
+
+@final
+class D1(TypedDict):
+    foo: int
+
+
+@final
+class D2(TypedDict):
+    bar: int
+
+d_or_list: D1 | list[str]
+
+if 'foo' in d_or_list:
+    assert_type(d_or_list, Union[D1, list[str]])
+else:
+    assert_type(d_or_list, list[str])
+
+d: D1 | D2
+
+if 'foo' in d:
+    assert_type(d, D1)
+else:
+    assert_type(d, D2)
+
+foo_or_bar: Literal['foo', 'bar']
+if foo_or_bar in d:
+    assert_type(d, Union[D1, D2])
+else:
+    assert_type(d, Union[D1, D2])
+
+foo_or_invalid: Literal['foo', 'invalid']
+if foo_or_invalid in d:
+    assert_type(d, D1)
```

Comment:
That'd have to be implemented.

Reply:
My current implementation is to make a form of "tagged TypeDicts" work, but I suspect a more generalized form of type narrowing should be possible. However, before we tackle the more contrived `x in y` case, we should probably do it for `x == y`, so that [this example](https://mypy-play.net/?mypy=latest&python=3.10&gist=18bfd81a2d803d34916d5e2b64b2247c) would see narrowing applied.

Reply:
Oh sorry, to be clear my suggestion wasn't to implement this / I don't think it's a terribly important feature. I was just saying it's worth adding an `assert_type` in this test to make this (lack of) behaviour clear to the reader.

Reply:
Gotcha, done: fb564eb75bbb4a76be08c6763a5b8da69bc724ee